### ISSUE TITLE
Questions about items produced by dataset's __getitem__() for multi-target multi-series multi-horizon predictions

### ISSUE THREAD

**Issue by maxkazmsft** [unit8co/darts#918]
I have a dataset which consists of multiple time series of varying length - you can think of each time series as a recording from sensors on some device for some time. The time index is equally-spaced, say it's every second, and has no real meaning beyond that for each device. Each recording is of varying length. Say I want to use the TCN model to predict sensor values on a new recording.

Each series has `nlag by nf` auto-regressive features, `nlag by nc` past covariates and I want to predict `npred by nf` sequence. Say `nlag=10`, `nf=4`, `nc=2`, `npred=5` and I have `N=100` such series.

When I look at the items produced by `GenericShiftedDataset` just before the call to `self.fit_from_dataset` in `torch_forecasting_model.py`, items from `train_dataset.__getitem__()` look strange:

1.  First, the items are mostly shifted by one time step, but in `GenericShiftedDataset` they should be shifted by `input_chunk_length`, which in this case is `nlag=10`.
2. In the tuple produced by `__getitem__()`, last element has dimensions `nlag by nf` instead of `npred by nf` and first `nlag-npred` rows overlaps with last `nlag-npred` rows from the first tuple element. I would expect the last element to just be a time continuation of the first element for `npred` steps, thus forming an `npred by nf` array of future values we're trying to predict.
3. The number of items quoted by `__len__()` doesn't make any sense either - if the series is shifted by 1, and there the M observations, then there should be `M-npred-nlag` items for each series, but this is not the case.
4. Some entries produced by `__getitem()__` are duplicates for some sequential `__getitem__(i)` and `__getitem__(i-1)` indices `i`.

I have sample code which I can share which illustrates these issues - how can I share it? Am I doing something wrong with the time index of each series (right now it's just an integer counter from 0 to the length of each series for each series)?

**Comment by hrzn** [comment]
Hi,

I'm not sure exactly what `GenericDataset` you are inspecting, because this is mostly a helper class that is not directly instantiated by the models. Rather, if you call `fit()` on say, a TCNModel, this model will instantiate for you a `PastCovariatesSequentialDataset`, which [internally makes use of a `GenericDataset`](https://github.com/unit8co/darts/blob/e3f33094ba0b02a96b25db9101b2be6def2d9fc7/darts/utils/data/sequential_dataset.py#L72). Let me try to answer your questions assuming you're using such a `PastCovariatesSequentialDataset`.

> 1. First, the items are mostly shifted by one time step, but in `GenericShiftedDataset` they should be shifted by `input_chunk_length`, which in this case is `nlag=10`.

The `GenericShiftedDataset` is instantiated with `shift=input_chunk_length`, which corresponds to you `nlag`. So the "future" item in the returned tuples should come `nlag` after the start of the "past" items. The different samples returned by different calls to `__getitem__()`, though, are typically separated by one time step. So for instance if you have a single training series with values `[1,2,3,4,5,6,7,8,9]` and `nlag=3` and `output_chunk_length=2`, the emmited training samples would be something like:
* 0-th training sample: `([1,2,3], [4,5])`,
* 1st training sample: `([2,3,4], [5,6])`,
etc...

> 2. In the tuple produced by `__getitem__()`, last element has dimensions `nlag by nf` instead of `npred by nf`

This should not be the case. Could you send a code snippet to reproduce this?

> and first `nlag-npred` rows overlaps with last `nlag-npred` rows from the first tuple element. I would expect the last element to just be a time continuation of the first element for `npred` steps, thus forming an `npred by nf` array of future values we're trying to predict.

I hope my answer on point 1 answers this.

> 3. The number of items quoted by `__len__()` doesn't make any sense either - if the series is shifted by 1, and there the M observations, then there should be `M-npred-nlag` items for each series, but this is not the case.

Your computation is more or less correct, except that `M` is taken as the maximum length of all your series. See the corresponding explanation in the dataset docstring:
```
The sampling is uniform over the number of time series; i.e., the i-th sample of this dataset has a probability 1/N of coming from any of the N time series in the sequence. If the time series have different lengths, they will contain different numbers of slices. Therefore, some particular slices may be sampled more often than others if they belong to shorter time series.
```

> 4. Some entries produced by `__getitem()__` are duplicates for some sequential `__getitem__(i)` and `__getitem__(i-1)` indices `i`.

I suspect this might be due to the behaviour I described just above. In case you want a different way to draw the samples, you can write your own implementation of `TrainingDataset`. Having a version that does not "force" the probably of 1/N to come from each of the N series for each sample might also have some general interest, so if you have such a dataset implementation you can PR it ;)

> 
> I have sample code which I can share which illustrates these issues - how can I share it? Am I doing something wrong with the time index of each series (right now it's just an integer counter from 0 to the length of each series for each series)?

This should not cause any issue.

I hope this helps.

**Comment by maxkazmsft** [comment]
Thanks, I pushed this script which illustrates the issues https://github.com/maxkazmsft/darts/blob/maxkazmsft/test_tcn/examples/05-TCN-series-loader.py

Please take a look and let me know what you think. I have another script which bypasses the `.fit` and calls `.fit_from_dataset` directly - if the example above isn't clear I can also show you the issues through that script too, but the above should be sufficient.

Also, I'm happy to hop on a quick call to step through the code if you have a channel for that already.

**Comment by hrzn** [comment]
Hi, looking at your example, the dataset will be instantiated as follows within the model:
```
ds = GenericShiftedDataset(
    target_series=train_series,
    covariates=train_series_cov,
    input_chunk_length=nlag,
    output_chunk_length=npred,
    shift=nlag,
    shift_covariates=False,
    max_samples_per_ts=None,
    covariate_type=CovariateType.PAST,
)
```

and when you look at e.g. `ds[0]` this is what you'd get:
```
(array([[0., 0., 0., 0.],
        [1., 1., 1., 1.],
        [2., 2., 2., 2.],
        [3., 3., 3., 3.],
        [4., 4., 4., 4.],
        [5., 5., 5., 5.],
        [6., 6., 6., 6.],
        [7., 7., 7., 7.],
        [8., 8., 8., 8.],
        [9., 9., 9., 9.]], dtype=float32),
 array([[0., 0.],
        [1., 1.],
        [2., 2.],
        [3., 3.],
        [4., 4.],
        [5., 5.],
        [6., 6.],
        [7., 7.],
        [8., 8.],
        [9., 9.]], dtype=float32),
 array([[10., 10., 10., 10.],
        [11., 11., 11., 11.],
        [12., 12., 12., 12.],
        [13., 13., 13., 13.],
        [14., 14., 14., 14.]], dtype=float32))
```

which to me looks good - e.g. the dependent variable has dimension `npred x nf` as expected. I guess you might have been inspecting the wrong dataset?

For the record, here's the code I used to get the output above:
```python
import numpy as np
import torch
from darts import TimeSeries
from pytorch_lightning import Trainer
from darts.utils.data.shifted_dataset import GenericShiftedDataset
from darts.utils.data.utils import CovariateType

# for reproducibility
torch.manual_seed(1)
np.random.seed(1)

from darts.models import (
    RNNModel,
    TCNModel,
)

'''
params
'''
nlag = 10
npred = 5
nf=4
nc=2


'''
create the actual dataset - just numbered rows for debugging
'''
def make_array(shape_tuple):

    array = np.ones(shape=shape_tuple).T
    for i in range(0,shape_tuple[0]):
        array[:,i]*=i

    return array.T


train_series = []
train_series_cov = []
nrow = nlag+npred

train_series.append(TimeSeries.from_values(make_array((nrow, nf))).astype(np.float32))
train_series_cov.append(TimeSeries.from_values(make_array((nrow, nc))).astype(np.float32))

# offset second series by 100 to tell the difference in debugger
train_series.append(TimeSeries.from_values(make_array((nrow+1, nf))+100).astype(np.float32))
train_series_cov.append(TimeSeries.from_values(make_array((nrow+1, nc))+100).astype(np.float32))

ds = GenericShiftedDataset(
    target_series=train_series,
    covariates=train_series_cov,
    input_chunk_length=nlag,
    output_chunk_length=npred,
    shift=nlag,
    shift_covariates=False,
    max_samples_per_ts=None,
    covariate_type=CovariateType.PAST,
)

ds[0]
```

**Comment by maxkazmsft** [comment]
Thanks, this still doesn't solve the issue - I updated the code here with more info
https://github.com/maxkazmsft/darts/blob/maxkazmsft/test_tcn/examples/05-TCN-series-loader.py

There seem to be two separate problems:
1) when I call `model.fit`, it eventually calls line 72 in `darts/utils/data/shifted_dataset.py` which has `output_chunk_length=nlag` (instead of `npred`) and `shift=npred` (instead of `nlag`). Even if I set `output_chunk_length=npred` and `shift=nlag` like you have in your example, I still get a stride of 1 and an incorrect number of `__len__()` (4 datapoints) - please see the new updated comments in my code.
2) when also calling `model.fit`, place a breakpoint at line 771 in `darts/models/forecasting/torch_forecasting_model.py` and check the data dimensions returned by `train_dataset.__getitem__()` - the dependent variable has the same problem `nlag by nif` dimensionality (instead of `npred by nif`) with `train_dataset.__len__()` function and also producing the wrong number of outputs.

In other words, to reproduce the issues, run my script and place breakpoints at 
- line 72 in `darts/utils/data/shifted_dataset.py` and inspect values set to `output_chunk_length` and `shift` variables
- line 771 in `darts/models/forecasting/torch_forecasting_model.py` and look at `train_dataset.__getitem__()` and `train_dataset.__len__()`

**Comment by hrzn** [comment]
Hi, OK, I think I understand. My previous reply was for the for the "general" past covariates case and I overlooked the fact that in your case you are using a `TCNModel`. The `TCNModel` [works differently](https://github.com/unit8co/darts/blob/e3f33094ba0b02a96b25db9101b2be6def2d9fc7/darts/models/forecasting/tcn_model.py#L474) and it using a "shifted" dataset instead of a "sequential" dataset. It is also outputting sequences of length `nlag`, because that's the way [TCN model needs to be trained](https://medium.com/unit8-machine-learning-publication/temporal-convolutional-networks-and-forecasting-5ce1b6e97ce4). When forecasting, the model outputs whole sequences of length `nlag`, but keeps only the last `npred` points of them (this is [defined here](https://github.com/unit8co/darts/blob/e3f33094ba0b02a96b25db9101b2be6def2d9fc7/darts/models/forecasting/tcn_model.py#L252)).
Hope this clarifies things.

**Comment by maxkazmsft** [comment]
OK, I understand how the TCN model is implemented with the shape of the dependent variable.

The issue still persists regarding the number of training datapoints. Here is how TCN model creates the training set:
```
ds = GenericShiftedDataset(
    target_series=train_series,
    covariates=train_series_cov,
    input_chunk_length=nlag,
    output_chunk_length=nlag,
    shift=npred,
    shift_covariates=False,
    max_samples_per_ts=None,
    covariate_type=CovariateType.PAST,
)
```
and the output is
```
ds.__len__()
Out[3]: 4
ds[0]
Out[4]: 
(array([[0., 0., 0., 0.],
        [1., 1., 1., 1.],
        [2., 2., 2., 2.],
        [3., 3., 3., 3.],
        [4., 4., 4., 4.],
        [5., 5., 5., 5.],
        [6., 6., 6., 6.],
        [7., 7., 7., 7.],
        [8., 8., 8., 8.],
        [9., 9., 9., 9.]], dtype=float32),
 array([[0., 0.],
        [1., 1.],
        [2., 2.],
        [3., 3.],
        [4., 4.],
        [5., 5.],
        [6., 6.],
        [7., 7.],
        [8., 8.],
        [9., 9.]], dtype=float32),
 array([[ 5.,  5.,  5.,  5.],
        [ 6.,  6.,  6.,  6.],
        [ 7.,  7.,  7.,  7.],
        [ 8.,  8.,  8.,  8.],
        [ 9.,  9.,  9.,  9.],
        [10., 10., 10., 10.],
        [11., 11., 11., 11.],
        [12., 12., 12., 12.],
        [13., 13., 13., 13.],
        [14., 14., 14., 14.]], dtype=float32))
ds[1]
Out[5]: 
(array([[0., 0., 0., 0.],
        [1., 1., 1., 1.],
        [2., 2., 2., 2.],
        [3., 3., 3., 3.],
        [4., 4., 4., 4.],
        [5., 5., 5., 5.],
        [6., 6., 6., 6.],
        [7., 7., 7., 7.],
        [8., 8., 8., 8.],
        [9., 9., 9., 9.]], dtype=float32),
 array([[0., 0.],
        [1., 1.],
        [2., 2.],
        [3., 3.],
        [4., 4.],
        [5., 5.],
        [6., 6.],
        [7., 7.],
        [8., 8.],
        [9., 9.]], dtype=float32),
 array([[ 5.,  5.,  5.,  5.],
        [ 6.,  6.,  6.,  6.],
        [ 7.,  7.,  7.,  7.],
        [ 8.,  8.,  8.,  8.],
        [ 9.,  9.,  9.,  9.],
        [10., 10., 10., 10.],
        [11., 11., 11., 11.],
        [12., 12., 12., 12.],
        [13., 13., 13., 13.],
        [14., 14., 14., 14.]], dtype=float32))
ds[2]
Out[6]: 
(array([[101., 101., 101., 101.],
        [102., 102., 102., 102.],
        [103., 103., 103., 103.],
        [104., 104., 104., 104.],
        [105., 105., 105., 105.],
        [106., 106., 106., 106.],
        [107., 107., 107., 107.],
        [108., 108., 108., 108.],
        [109., 109., 109., 109.],
        [110., 110., 110., 110.]], dtype=float32),
 array([[101., 101.],
        [102., 102.],
        [103., 103.],
        [104., 104.],
        [105., 105.],
        [106., 106.],
        [107., 107.],
        [108., 108.],
        [109., 109.],
        [110., 110.]], dtype=float32),
 array([[106., 106., 106., 106.],
        [107., 107., 107., 107.],
        [108., 108., 108., 108.],
        [109., 109., 109., 109.],
        [110., 110., 110., 110.],
        [111., 111., 111., 111.],
        [112., 112., 112., 112.],
        [113., 113., 113., 113.],
        [114., 114., 114., 114.],
        [115., 115., 115., 115.]], dtype=float32))
ds[3]
Out[7]: 
(array([[100., 100., 100., 100.],
        [101., 101., 101., 101.],
        [102., 102., 102., 102.],
        [103., 103., 103., 103.],
        [104., 104., 104., 104.],
        [105., 105., 105., 105.],
        [106., 106., 106., 106.],
        [107., 107., 107., 107.],
        [108., 108., 108., 108.],
        [109., 109., 109., 109.]], dtype=float32),
 array([[100., 100.],
        [101., 101.],
        [102., 102.],
        [103., 103.],
        [104., 104.],
        [105., 105.],
        [106., 106.],
        [107., 107.],
        [108., 108.],
        [109., 109.]], dtype=float32),
 array([[105., 105., 105., 105.],
        [106., 106., 106., 106.],
        [107., 107., 107., 107.],
        [108., 108., 108., 108.],
        [109., 109., 109., 109.],
        [110., 110., 110., 110.],
        [111., 111., 111., 111.],
        [112., 112., 112., 112.],
        [113., 113., 113., 113.],
        [114., 114., 114., 114.]], dtype=float32))
```

I expect the output to be 
```
ds.__len__()
Out[3]: 2
ds[0]
Out[4]: 
(array([[0., 0., 0., 0.],
        [1., 1., 1., 1.],
        [2., 2., 2., 2.],
        [3., 3., 3., 3.],
        [4., 4., 4., 4.],
        [5., 5., 5., 5.],
        [6., 6., 6., 6.],
        [7., 7., 7., 7.],
        [8., 8., 8., 8.],
        [9., 9., 9., 9.]], dtype=float32),
 array([[0., 0.],
        [1., 1.],
        [2., 2.],
        [3., 3.],
        [4., 4.],
        [5., 5.],
        [6., 6.],
        [7., 7.],
        [8., 8.],
        [9., 9.]], dtype=float32),
 array([[ 5.,  5.,  5.,  5.],
        [ 6.,  6.,  6.,  6.],
        [ 7.,  7.,  7.,  7.],
        [ 8.,  8.,  8.,  8.],
        [ 9.,  9.,  9.,  9.],
        [10., 10., 10., 10.],
        [11., 11., 11., 11.],
        [12., 12., 12., 12.],
        [13., 13., 13., 13.],
        [14., 14., 14., 14.]], dtype=float32))
ds[1]
Out[5]: 
(array([[100., 100., 100., 100.],
        [101., 101., 101., 101.],
        [102., 102., 102., 102.],
        [103., 103., 103., 103.],
        [104., 104., 104., 104.],
        [105., 105., 105., 105.],
        [106., 106., 106., 106.],
        [107., 107., 107., 107.],
        [108., 108., 108., 108.],
        [109., 109., 109., 109.]], dtype=float32),
 array([[100., 100.],
        [101., 101.],
        [102., 102.],
        [103., 103.],
        [104., 104.],
        [105., 105.],
        [106., 106.],
        [107., 107.],
        [108., 108.],
        [109., 109.]], dtype=float32),
 array([[105., 105., 105., 105.],
        [106., 106., 106., 106.],
        [107., 107., 107., 107.],
        [108., 108., 108., 108.],
        [109., 109., 109., 109.],
        [110., 110., 110., 110.],
        [111., 111., 111., 111.],
        [112., 112., 112., 112.],
        [113., 113., 113., 113.],
        [114., 114., 114., 114.]], dtype=float32))
```

Does the actual output and expected output above make sense?

I updated by sample script with the above again.

**Comment by hrzn** [comment]
Your two training series have length 15 and 16, respectively. Here, the number of training samples per time series is determined by the longest series - i.e., the one of length 16. This series maps to 2 training samples:
* One training sample covering `100 -> 109` (input) and `105 -> 114` (output)
* Another training sample covering `101 -> 110` (input) and `106 -> 115` (output)
The first training series of length 15 will also be mapped to 2 training samples. However, because it is of length 15, those 2 training samples are the same (covering `0 -> 9` and `5 -> 14`).

This matches the description of the datasets. For instance for [PastCovariatesShiftedDataset](https://unit8co.github.io/darts/generated_api/darts.utils.data.shifted_dataset.html#darts.utils.data.shifted_dataset.PastCovariatesShiftedDataset) used in the TCNModel:
```
The sampling is uniform over the number of time series; i.e., the i-th sample of this dataset has a probability 1/N of coming from any of the N time series in the sequence. If the time series have different lengths, they will contain different numbers of slices. Therefore, some particular slices may be sampled more often than others if they belong to shorter time series.
```
So here there are 4 samples, each having probability 1/2 of coming from the first series, and 1/2 of coming from the second series. Since the first series is shorter, it contains slices that are "sampled more often than others".

If you want/need to change this behaviour, you can for instance specify `max_samples_per_ts=1` when creating the dataset, or implement another `TrainingDataset` yourself which implements the slicing logic you want.

**Comment by maxkazmsft** [comment]
Thank you for clarifying! If you set `shift=1` above, then what you said makes perfect sense and I understand the sampling strategy - all good.

However, `shift=npred`, so I would expect the first series to generate 1 training sample, then the second series to also generate just 1 training sample and stop there, because it cannot move the sliding window by `shift>1` points (it'll run out of data). The only way the produced output makes sense is if the sliding window moves from right to left in the second series, and includes the left-most position of the sliding window if the series runs out of data. Does this make sense? Then second series will produce 2 points and then first series will get sampled twice for a total of 4 points like in the output.

PS: feature ask - make `max_samples_per_ts` take a list of integers, where each integer specifies max number of samples from each time series.

I'll try implementing my own `TrainingDataset` and might post this as a separate issue and cross-reference here.

**Comment by hrzn** [comment]
I think there might be a misunderstanding. What you describe as `shift` sounds like a "stride" - by how much you slide a certain window. In all datasets implemented here, this is always 1. So that's why your second series of length 16 yields 2 training samples. In contrast, the `shift` parameters refers to: `The number of time steps by which to shift the output chunks relative to the input chunks.`; so basically if `shift=5` and the input chunk starts at time `t`, then the output chunks will start at time `t+5` - I hope this clarifies it.

If you implement your own `TrainingDataset` and you suspect it might be of interest, we would happily take a PR around this :)

**Comment by maxkazmsft** [comment]
Thanks, this is sufficient to close the issue - I'll follow up with a script which trains the model with `.fit_from_dataset` (where I already noticed that to call `model.predict` one needs to set `model._fit_called=True` first after training the model.
